src/ngx_http_lua_uri.c

```diff
@@ -107,4 +110,38 @@ ngx_http_lua_ngx_req_set_uri(lua_State *L)
     return 0;
 }
 
+
+static ngx_inline size_t
+ngx_http_lua_safe_uri_len(u_char *str, size_t len)
+{
+    size_t i;
+
+                    /* %00-%1F, " ", %7F */
+
+    static uint32_t unsafe[] = {
+        0xffffffff, /* 1111 1111 1111 1111  1111 1111 1111 1111 */
+
+                    /* ?>=< ;:98 7654 3210  /.-, +*)( '&%$ #"!  */
+        0x00000001, /* 0000 0000 0000 0000  0000 0000 0000 0001 */
+
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+
+                    /*  ~}| {zyx wvut srqp  onml kjih gfed cba` */
+        0x80000000, /* 1000 0000 0000 0000  0000 0000 0000 0000 */
+
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000  /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+    };
+
+    for (i = 0; i < len; i++, str++) {
+        if (unsafe[*str >> 5] & (1 << (*str & 0x1f))) {
+            return i;
```

Comment:
I think it's safer to return `(size_t) -1` here to indicate an error instead of do blind truncation. Blind truncation may still have security implications.

Reply:
Also need debug logging here to show the offensive byte.